python/cuml/tsa/arima.pyx

```diff
@@ -863,7 +844,65 @@ class ARIMA(Base):
     def llf(self):
         """Log-likelihood of a fit model. Shape: (batch_size,)
         """
-        return self._loglike(self.pack(), trans=False)
+        # Implementation note: this is slightly different from batched_loglike
+        # as it uses the device parameter arrays and not a host vector.
+        # Also, it always uses the MLE method, trans=False and truncate=0
+
+        cdef handle_t* handle_ = <handle_t*><size_t>self.handle.getHandle()
+
+        cdef vector[double] vec_loglike
+        vec_loglike.resize(self.batch_size)
+
+        cdef ARIMAOrder order_kf = \
+            self.order_diff if self.simple_differencing else self.order
+
+        # Convert host parameters to device parameters
+        cdef uintptr_t d_mu_ptr = <uintptr_t> NULL
+        cdef uintptr_t d_ar_ptr = <uintptr_t> NULL
+        cdef uintptr_t d_ma_ptr = <uintptr_t> NULL
+        cdef uintptr_t d_sar_ptr = <uintptr_t> NULL
+        cdef uintptr_t d_sma_ptr = <uintptr_t> NULL
+        cdef uintptr_t d_sigma2_ptr = <uintptr_t> NULL
+        if order_kf.k:
+            d_mu_ptr = self._mu_.ptr
+        if order_kf.p:
+            d_ar_ptr = self._ar_.ptr
+        if order_kf.q:
+            d_ma_ptr = self._ma_.ptr
+        if order_kf.P:
+            d_sar_ptr = self._sar_.ptr
+        if order_kf.Q:
+            d_sma_ptr = self._sma_.ptr
+        d_sigma2_ptr = self._sigma2_.ptr
+
+        cdef ARIMAParams[double] cpp_params
+        cpp_params.mu = <double*> d_mu_ptr
+        cpp_params.ar = <double*> d_ar_ptr
+        cpp_params.ma = <double*> d_ma_ptr
+        cpp_params.sar = <double*> d_sar_ptr
+        cpp_params.sma = <double*> d_sma_ptr
+        cpp_params.sigma2 = <double*> d_sigma2_ptr
+
+        cdef uintptr_t d_y_kf_ptr = \
+            self._d_y_diff.ptr if self.simple_differencing else self._d_y.ptr
+
+        n_obs_kf = (self.n_obs_diff if self.simple_differencing
+                    else self.n_obs)
+
+        cdef LoglikeMethod ll_method = MLE
+        diff = self.simple_differencing
+
+        d_vs = cumlArray.empty((n_obs_kf, self.batch_size), dtype=np.float64,
+                               order="F")
+        cdef uintptr_t d_vs_ptr = d_vs.ptr
+
+        batched_loglike(handle_[0], <double*> d_y_kf_ptr,
+                        <int> self.batch_size, <int> n_obs_kf, order_kf,
+                        cpp_params, <double*> vec_loglike.data(),
+                        <double*> d_vs_ptr, <bool> False, <bool> True,
+                        ll_method, <int> 0)
+
+        return np.array(vec_loglike, dtype=np.float64)
```

Comment:
Unclear why `llf` changed so much here. Whats the reasoning for not using `self._loglike()` anymore?

Reply:
`_loglike()` and `_loglike_grad()` are primarily used by the scipy interpreter so they take a numpy parameter array, which is then copied to device and unpacked. `llf` is using the final parameters which are now stored as cuML arrays, so already unpacked and on device. It wouldn't make sense to copy the parameters from device to host then device again.

What I could do to remove redundancy, however, is to create a third function similar to the current `llf`, that takes the unpacked device arrays, and base both functions on this third one by moving the unpacking of `_batched_loglike` to the Python side. But this won't really be much more elegant than the current solution, and certainly not faster, so at this point I would just leave it like this, and the redundancy will disappear when this code moves to the C++ side.

Reply:
Not sure i followed all of that, but I'll take your word for it.